### PR TITLE
Don't generate nullable type when applying value type with unconstrained nullable generic

### LINKED ISSUE #48742: Implement abstract class incorrect for unconstrained generic nullable in base and value type in derived
**Version Used**: VS 2019 version 16.8.0 Preview 4.0

**Steps to Reproduce**:

1. Have code like the following:

    ```c#
    abstract class B<T>
    {
        public abstract T? M();
    }

    class D : B<int>
    {
    }
    ```
2. Use <kbd>Ctrl</kbd>+<kbd>.</kbd> to select "implement abstract class".

**Expected Behavior**:

This generates code that compiles, i.e.:

```c#
public override int M()
{
    throw new NotImplementedException();
}
```

**Actual Behavior**:

The generated override uses a nullable value type, which doesn't compile:

```c#
public override int? M()
{
    throw new NotImplementedException();
}
```

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Fixes #48742. Should prevent similar issue in other cases.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/ImplementAbstractClass/ImplementAbstractClassTests.cs [hunk 1]
```diff
@@ -1911,5 +1911,59 @@ public override void AbstractMethod()
...
+}
+
+class D<T> : B<T> where T : struct
+{
+    public override T M()
+    {
+        throw new System.NotImplementedException();
+    }
+}");
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ImplementAbstractClass/ImplementAbstractClassTests.cs:1967]: can we a few more tests.  

1. where T is unconstrained
2. where T is constrained to `class`.  
3. where T is unconstrained, and passed as T? to the base type
3. where T is constrained to `class`, and passed as T? to the base type
3. where T is constrained to `struct`, and passed as T? to the base type

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs [hunk 2]
```diff
@@ -300,8 +300,10 @@ public override TypeSyntax VisitNamedType(INamedTypeSymbol symbol)
                     }
                 }
 
-                if (symbol.NullableAnnotation == NullableAnnotation.Annotated)
+                if (symbol is { IsValueType: false, NullableAnnotation: NullableAnnotation.Annotated })
                 {
+                    // value type with nullable annotation may be composed from unconstrained nullable generic
+                    // doesn't mean nullable value type in this case
                     typeSyntax = AddInformationTo(SyntaxFactory.NullableType(typeSyntax), symbol);
                 }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:308]: what about other named types that can be value types.  like tuples.  can you update as well (or add tests showing they're k).

File: src/EditorFeatures/CSharpTest/ImplementAbstractClass/ImplementAbstractClassTests.cs (hunk 1 above)
Comment (huoyaoyuan) [reply src/EditorFeatures/CSharpTest/ImplementAbstractClass/ImplementAbstractClassTests.cs:1967]: Sure, make sure that other cases doesn't regress.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs (hunk 2 above)
Comment (huoyaoyuan) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:308]: Are they special for this case? Here is adding top-level nullability, I couldn't imagine any type to have special semantics. The inner type has been handled above.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:308]: they're a named type that is a value type :)  i just want to make sure it works properly and nothing unexpected happens.

Comment (huoyaoyuan) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:308]: Adding tests should be OK I think

**Comment by huoyaoyuan** [comment]
Well, this is actually duplicated by #49242 . The corresponding issue can be closed now.

**Comment by huoyaoyuan** [comment]
@CyrusNajmabadi this has more test coverage than #49242 . Idea?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi this has more test coverage than #49242 . Idea?

You can reopen and just include the tests you want to add. Thanks!

**Comment by huoyaoyuan** [comment]
I think it's better to keep this PR, because it adds more comment and addresses another position.

**Comment by huoyaoyuan** [comment]
@CyrusNajmabadi Idea on this? Do you think it's worth to keep it?

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**